README.md
virtualenv-api - an API for virtualenv
======================================

[![Build Status](https://travis-ci.org/sjkingo/virtualenv-api.svg)](https://travis-ci.org/sjkingo/virtualenv-api)
[![Latest version](https://img.shields.io/pypi/v/virtualenv-api.svg)](https://pypi.python.org/pypi/virtualenv-api)
[![BSD License](https://img.shields.io/pypi/l/virtualenv-api.svg)](https://github.com/sjkingo/virtualenv-api/blob/master/LICENSE)
[![Stories in Ready](https://badge.waffle.io/sjkingo/virtualenv-api.png?label=ready&title=Ready)](https://waffle.io/sjkingo/virtualenv-api)

[virtualenv](http://www.virtualenv.org/) is a tool to create isolated Python
environments.  Unfortunately, it does not expose a native Python API. This
package aims to provide an API in the form of a wrapper around virtualenv.

It can be used to create and delete environments and perform package management
inside the environment.

Full support is provided for Python 2.7 and Python 3.3+.

Installation
------------

The latest stable release is available on [PyPi](https://pypi.python.org/pypi/virtualenv-api):

```
$ pip install virtualenv-api
```

Please note that the distribution is named `virtualenv-api`, yet the Python package
is named `virtualenvapi`.

Alternatively, you may fetch the latest version from git:

```
$ pip install git+https://github.com/sjkingo/virtualenv-api.git
```

Examples
--------

* To begin managing an environment (it will be created if it does not exist):

```python
from virtualenvapi.manage import VirtualEnvironment
env = VirtualEnvironment('/path/to/environment/name')
```

You may also specify the Python interpreter to use in this environment by
passing the `python` argument to the class constructor (new in 2.1.3):

```python
env = VirtualEnvironment('/path/to/environment/name', python='python3')
```

If you have already activated a virtualenv and wish to operate on it, simply call
`VirtualEnvironment` with no arguments:

```python
env = VirtualEnvironment()
```

*New in 2.1.7:*

An optional argument `readonly` may be provided (defaults to `False`) that 
will prevent all operations that could potentially modify the environment.

* Check if the `mezzanine` package is installed:

```python
>>> env.is_installed('mezzanine')
False
```

* Install the latest version of the `mezzanine` package:

```python
>>> env.install('mezzanine')
```

* A wheel of the latest version of the `mezzanine` package (new in 2.1.4):

```python
>>> env.wheel('mezzanine')
```

* Install version 1.4 of the `django` package (this is pip's syntax):

```python
>>> env.install('django==1.4')
```

* Upgrade the `django` package to the latest version:

```python
>>> env.upgrade('django')
```

* Upgrade all packages to their latest versions (new in 2.1.7):

```python
>>> env.upgrade_all()
```

* Uninstall the `mezzanine` package:

```python
>>> env.uninstall('mezzanine')
```

Packages may be specified as name only (to work on the latest version),
using pip's package syntax (e.g. `django==1.4`) or as a tuple of `('name', 'ver')`
(e.g. `('django', '1.4')`).

* A package may be installed directly from a git repository (must end with `.git`):

```python
>>> env.install('git+git://github.com/sjkingo/cartridge-payments.git')
```

* Instances of the environment provide an `installed_packages` property:

```python
>>> env.installed_packages
[('django', '1.5'), ('wsgiref', '0.1.2')]
```

* A list of package names is also available in the same manner:

```python
>>> env.installed_package_names
['django', 'wsgiref']
```

* Search for a package on PyPI (changed in 2.1.5: this now returns a dictionary instead of list):

```python
>>> env.search('virtualenv-api')
{'virtualenv-api': 'An API for virtualenv/pip'}
>>> len(env.search('requests'))
231
```

* The old functionality (pre 2.1.5) of `env.search` may be used:

```python
>>> list(env.search('requests').items())
[('virtualenv-api', 'An API for virtualenv/pip')]
```

Verbose output from each command is available in the environment's `build.log`
file, which is appended to with each operation. Any errors are logged to `build.err`.


CHANGES.md
# Changes

## 2.1.18 - 2020-02-03

* #46: Blacklist env var in subprocess calls to fix bug in MacOS/Homebrew installs (@irvinlim)
* Change build versions to match supported Python versions (@sjkingo)

## 2.1.17 - 2018-11-15

* #40: Fix for test cases failing (@irvinlim)
* #39: Fix for `is_installed` not handling capitilized names properly (@irvinlim)
* Support Python 3.7 (@sjkingo)

## 2.1.16 - 2017-03-17

* Fix bug in test script where temporary directories were not being removed (@sjkingo)
* #35: Don't pass unsupported argument to old versions of pip (@sjkingo)
* Always call pip through the Python interpreter to avoid a too-long shebang error (@sjkingo)
* #26: Fix incorrect subclassing of base test class that was causing extreme test durations (@sjkingo)

## 2.1.15 - 2017-03-17

* #34: Add support for `pip -r` requirements files (@jzafran and @sjkingo)

## 2.1.14 - 2017-02-22

* #33: Add support for `system-site-packages` option (@evansde77)

## 2.1.13 - 2016-10-26

* #31: Workaround to prevent shebang length errors when calling pip (@rmb938)

## 2.1.12 - 2016-10-22

* #29: Fix AttributeError when raising OSError from inside environment (@rmb938)

## 2.1.11 - 2016-07-14

* #28: Add support for locating pip on win32 (@eoghancunneen)

## 2.1.10 - 2016-06-03

* #27: Support installing an editable package (`-e`) (@sjkingo)

## 2.1.9 - 2016-05-01

* Move version number to library instead of `setup.py` (@sjkingo)
* Support Python 3.5 (@sjkingo)
* #24: Fixed failing test suite by adding missing dependency (@mcyprian)

## 2.1.8 - 2016-04-01

* #21: Converted `README.md` to an rST document to remove dependency on
  `setuptools-markdown` (@sjkingo)

## 2.1.7 - 2015-08-10

* Added `upgrade_all()` function to upgrade all packages in an environment (@sjkingo)
* Added `readonly` argument to constructing environment that can be used to prevent
  operations that could potentially modify the environment (@sjkingo)
* Added support for passing ~ to construct environment (e.g. `~user/venv`) (@sjkingo)

## 2.1.6 - 2015-06-10

 * Version bump for broken PyPi release (@sjkingo)
   (no new changes to code)

## 2.1.5 - 2015-06-10

 * Improved search function that will return more accurate results. This
   includes a breaking change where the package list returned by `env.search()`
   is now a dictionary. (@sjkingo)
 * Prevent pip from checking for new version of itself and polluting the output
   of some commands (@sjkingo)

## 2.1.4 - 2015-06-04

 * Support for creating a wheel of packages (@rmb938)

## 2.1.3 - 2015-03-28

 * Support changing the interpreter path (@jlafon)
 * Improve Unicode support in pip search that broke tests (@jlafon)

## 2.1.2 - 2014-11-25

 * Added test builds through Travis CI (@sjkingo)
 * Fixed default `options` bug introduced in 2.0.1 (@ColMcp, @sposs)
 * Updated example.py (@sjkingo)
 * Fix typo in logging (@yannik-ammann)

## 2.1.1 - 2014-11-19

 * Fix typo that broke from 2.1.0 release (@sjkingo)

## 2.1.0 - 2014-11-19

 * Python 3 support (@r1s)
 * Unit tests for base functionality (@r1s)
 * Better Unicode handling (@r1s)
 * Tuple support for specifying package versions (@philippeowagner)

## 2.0.1 - 2014-11-14

 * Support passing command line options directly to pip (@sposs)
 * Misc. PEP8 fixes (@sposs)

## 2.0.0 - 2014-04-09

 * Added pip search functionality
 * Re-worked underlying pip processing and error handling

## 1.0.0 - 2013-03-27

 * Initial release with basic install/uninstall


test.py
#!/usr/bin/python

from virtualenv.manage import VirtualEnvironment

def sample_usage():
    env = VirtualEnvironment('/tmp/virtualenv.test')
    print 'mezzanine installed?', env.is_installed('mezzanine')
    env.install('mezzanine')
    print 'mezzanine installed?', env.is_installed('mezzanine')
    print 'django 1.5 installed?', env.is_installed('django==1.5')
    print env.installed_packages

    payments_repo = 'git+git://github.com/explodes/cartridge-payments.git'
    print 'cartridge-payments installed?', env.is_installed(payments_repo)
    env.install(payments_repo)
    print 'cartridge-payments installed?', env.is_installed(payments_repo)
    print env.installed_packages


if __name__ == '__main__':
    sample_usage()


tests.py
import os
import random
import shutil
import string
import sys
import tempfile
import unittest

from virtualenvapi.manage import VirtualEnvironment

packages_for_tests = ['pep8']
non_lowercase_packages_for_test = ['Pillow']
all_packages_for_tests = packages_for_tests + non_lowercase_packages_for_test

def which(program):
    def is_exe(fpath):
        return os.path.isfile(fpath) and os.access(fpath, os.X_OK)

    fpath, fname = os.path.split(program)

    if fpath:
        if is_exe(program):
            return program
    else:
        for path in os.environ["PATH"].split(os.pathsep):
            path = path.strip('"')
            exe_file = os.path.join(path, program)
            if is_exe(exe_file):
                return exe_file

    return None


class TestBase(unittest.TestCase):
    """
    Base class for test cases to inherit from.
    """

    env_path = None

    def setUp(self):
        self.env_path = tempfile.mkdtemp()
        self.virtual_env_obj = VirtualEnvironment(self.env_path)

    def tearDown(self):
        if os.path.exists(self.env_path):
            shutil.rmtree(self.env_path)

    def _install_packages(self, packages):
        for pack in packages:
            self.virtual_env_obj.install(pack)

    def _uninstall_packages(self, packages):
        for pack in packages:
            self.virtual_env_obj.uninstall(pack)


class InstalledTestCase(TestBase):
    """
    Test install/uninstall operations.
    """

    def test_installed(self):
        self.assertFalse(self.virtual_env_obj.is_installed(''.join(random.sample(string.ascii_letters, 30))))

    def test_install(self):
        for pack in all_packages_for_tests:
            self.virtual_env_obj.install(pack)
            self.assertTrue(self.virtual_env_obj.is_installed(pack))

    def test_install_requirements(self):
        """test installing Python packages from a pip requirements file"""

        # create a temporary requirements.txt file with some packages
        with tempfile.NamedTemporaryFile('w') as tmp_requirements_file:
            tmp_requirements_file.write('\n'.join(packages_for_tests))
            tmp_requirements_file.flush()

            self.virtual_env_obj.install('-r {}'.format(tmp_requirements_file.name))
            for pack in packages_for_tests:
                self.assertTrue(self.virtual_env_obj.is_installed(pack))

    def test_uninstall(self):
        self._install_packages(all_packages_for_tests)
        for pack in all_packages_for_tests:
            if pack.endswith('.git'):
                pack = pack.split('/')[-1].replace('.git', '')
            self.virtual_env_obj.uninstall(pack)
            self.assertFalse(self.virtual_env_obj.is_installed(pack))

    def test_wheel(self):
        self.virtual_env_obj.install('wheel') # required for this test
        for pack in packages_for_tests:
            self.virtual_env_obj.wheel(pack, options=['--wheel-dir=/tmp/wheelhouse'])
            self.virtual_env_obj.install(pack, options=['--no-index', '--find-links=/tmp/wheelhouse'])
            self.assertTrue(self.virtual_env_obj.is_installed(pack))


class SearchTestCase(TestBase):
    """
    Test pip search.
    """

    def test_search(self):
        for pack in all_packages_for_tests:
            result = self.virtual_env_obj.search(pack)
            self.assertIsInstance(result, dict)
            self.assertTrue(bool(result))
            if result:
                self.assertIn(pack.lower(), [k.split(' (')[0].lower() for k in result.keys()])

    def test_search_names(self):
        for pack in all_packages_for_tests:
            result = self.virtual_env_obj.search_names(pack)
            self.assertIsInstance(result, list)
            self.assertIn(pack.lower(), [k.split(' (')[0].lower() for k in result])


class PythonArgumentTestCase(TestBase):
    """
    Test passing a different interpreter path to `VirtualEnvironment` (`virtualenv -p`).
    """

    def setUp(self):
        self.env_path = tempfile.mkdtemp()
        self.python = which('python')
        self.assertIsNotNone(self.python)
        self.virtual_env_obj = VirtualEnvironment(self.env_path, python=self.python)

    def test_python_version(self):
        self.assertEqual(self.virtual_env_obj.python, self.python)
        self.assertEqual(
            os.path.dirname(self.python),
            os.path.dirname(which('pip'))
        )


class SystemSitePackagesTest(unittest.TestCase):
    """
    test coverage for using system-site-packages flag

    This verifies the presence of the no-global-site-packages
    file in the venv for non-system packages venv and
    verified it is not present for the true case

    """
    def setUp(self):
        """
        snapshot system package list
        """
        self.dir = tempfile.mkdtemp()
        self.no_global = (
            "lib/python{}.{}"
            "/no-global-site-packages.txt"
        ).format(
            sys.version_info.major,
            sys.version_info.minor
        )

    def tearDown(self):
        if os.path.exists(self.dir):
            shutil.rmtree(self.dir)

    def test_system_site_packages(self):
        """
        test that creating a venv with system_site_packages=True
        results in a venv that does not contain the no-global-site-packages file

        """
        venv = VirtualEnvironment(self.dir, system_site_packages=True)
        venv._create()
        expected = os.path.join(venv.path, self.no_global)
        self.assertTrue(
            not os.path.exists(expected)
        )

    def test_no_system_site_packages(self):
        """
        test that creating a venv with system_site_packages=False
        results in a venv that contains the no-global-site-packages
        file

        """
        venv = VirtualEnvironment(self.dir)
        venv._create()
        expected = os.path.join(venv.path, self.no_global)
        self.assertTrue(
            os.path.exists(expected)
        )


if __name__ == '__main__':
    unittest.main()


example.py
#!/usr/bin/env python
import tempfile

from virtualenvapi.manage import VirtualEnvironment


def example(path=tempfile.mkdtemp('virtualenv.test')):
    print('Env path', path)
    env = VirtualEnvironment(path)

    print('django 1.5 installed?', env.is_installed('django==1.5'))

    print('mezzanine installed?', env.is_installed('mezzanine'))
    env.install('mezzanine')
    print('mezzanine installed?', env.is_installed('mezzanine'))

    print(env.installed_packages)

    repo = 'git+git://github.com/sjkingo/django_auth_ldap3.git'
    pkg = repo.split('/')[-1].replace('.git', '')
    print('django_auth_ldap3 installed?', env.is_installed(pkg))
    env.install(repo)
    print('django_auth_ldap3 installed?', env.is_installed(pkg))
    print(env.installed_packages)

    env.uninstall('mezzanine')
    print('mezzanine installed?', env.is_installed('mezzanine'))
    print(env.installed_packages)

    pkgs = env.search('requests')
    print('search for \'requests\':')
    print(len(pkgs), 'found:')
    print(pkgs)


if __name__ == '__main__':
    example()


virtualenv.py
from os import linesep
import os.path
import re
import subprocess

class VirtualEnvironment(object):
    # True if the virtual environment has been set up through open_or_create()
    _ready = False

    def __init__(self, path):
        # remove trailing slash so os.path.split() behaves correctly
        if path[-1] == '/':
            path = path[:-1]
        self.path = path

    def __str__(self):
        return self.path

    @property
    def _pip_rpath(self):
        """The relative path (from environment root) to pip."""
        return os.path.join('bin', 'pip')

    @property
    def root(self):
        """The root directory that this virtual environment exists in."""
        return os.path.split(self.path)[0]

    @property
    def name(self):
        """The name of this virtual environment (taken from its path)."""
        return os.path.basename(self.path)

    @property
    def _logfile(self):
        """Absolute path of the log file for recording installation output."""
        return os.path.join(self.path, 'build.log')

    def _create(self):
        """Executes `virtualenv` to create a new environment."""
        out = subprocess.check_output(['virtualenv', self.name], cwd=self.root)
        self._write_to_log(out, truncate=True) # new log

    def _execute(self, args):
        """Executes the given command inside the environment and returns the output."""
        if not self._ready:
            self.open_or_create()
        return subprocess.check_output(args, cwd=self.path)

    def _write_to_log(self, s, truncate=False):
        """Writes the given output to the log file, appending unless `truncate` is True."""
        # if truncate is True, set write mode to truncate
        choice = { True: 'w',
                   False: 'a'
                 }
        with open(self._logfile, choice[truncate]) as fp:
            fp.write(s + linesep)

    def _pip_exists(self):
        """Returns True if pip exists inside the virtual environment. Can be
        used as a naive way to verify that the envrionment is installed."""
        return os.path.isfile(os.path.join(self.path, self._pip_rpath))

    def open_or_create(self):
        """Attempts to open the virtual environment or creates it if it
        doesn't exist.
        XXX this should probably be expanded to do some proper checking?"""
        if not self._pip_exists():
            self._create()
        self._ready = True

    def install(self, package, force=False):
        """Installs the given package (given in pip's package syntax) 
        into this virtual environment only if it is not already installed.
        If `force` is True, force an installation."""
        if not force and self.is_installed(package):
            self._write_to_log('%s is already installed, skipping (use force=True to override)' % package)
            return
        out = self._execute([self._pip_rpath, 'install', package])
        self._write_to_log(out)

    def is_installed(self, package):
        """Returns True if the given package (given in pip's package syntax)
        is installed in the virtual environment."""
        installed_packages = self._execute([self._pip_rpath, 'freeze']).lower()
        pkg_name = package.lower().split('=')[0]
        if pkg_name.endswith('.git'):
            pkg_name = os.path.split(pkg_name)[1][:-4]
        m = re.search(pkg_name, installed_packages)
        return True if m is not None else False

    def upgrade(self, package):
        """Shortcut method to upgrade a package by forcing a reinstall.
        Note that this may not actually upgrade but merely reinstall if there
        is no newer version to install."""
        self.install(package, force=True)


virtualenvapi/util.py
from os import environ
import six
import sys


def to_text(source):
    if six.PY3:
        if isinstance(source, str):
            return source
        else:
            return source.decode("utf-8")
    elif six.PY2:
        if isinstance(source, unicode):
            return source.encode("utf-8")
        return source
    else:
        return source


def to_ascii(source):
    if isinstance(source, six.string_types):
        return "".join([c for c in source if ord(c) < 128])


def get_env_path():
    prefix_name = 'real_prefix'
    virtual_env_path_environ_key = 'VIRTUAL_ENV'

    path = None

    real_prefix = (hasattr(sys, prefix_name) and getattr(sys, prefix_name)) or None
    if real_prefix:
        path = environ.get(virtual_env_path_environ_key)
        if not path:
            path = sys.prefix

    return path


def split_package_name(p):
    """Splits the given package name and returns a tuple (name, ver)."""
    s = p.split(six.u('=='))
    if len(s) == 1:
        return (to_text(s[0]), None)
    else:
        return (to_text(s[0]), to_text(s[1]))


virtualenvapi/manage.py
from os import linesep, environ
import os.path
import subprocess
import six
import sys

from virtualenvapi.util import split_package_name, to_text, get_env_path, to_ascii
from virtualenvapi.exceptions import *


class VirtualEnvironment(object):

    def __init__(self, path=None, python=None, cache=None, readonly=False, system_site_packages=False):

        if path is None:
            path = get_env_path()

        if not path:
            raise VirtualenvPathNotFound('Path for virtualenv is not define or virtualenv is not activate')

        self.python = python
        self.system_site_packages = system_site_packages

        # remove trailing slash so os.path.split() behaves correctly
        if path[-1] == os.path.sep:
            path = path[:-1]

        # Expand path so shell shortcuts may be used such as ~
        self.path = os.path.abspath(os.path.expanduser(path))

        self.env = environ.copy()

        # Blacklist environment variables that will break pip in virtualenvs
        # See https://github.com/pypa/virtualenv/issues/845
        self.env.pop('__PYVENV_LAUNCHER__', None)

        if cache is not None:
            self.env['PIP_DOWNLOAD_CACHE'] = os.path.expanduser(os.path.expandvars(cache))

        self.readonly = readonly

        # True if the virtual environment has been set up through open_or_create()
        self._ready = False

    def __str__(self):
        return six.u(self.path)

    @property
    def _pip(self):
        """The arguments used to call pip."""
        # pip is called using the python interpreter to get around a long path
        # issue detailed in https://github.com/sjkingo/virtualenv-api/issues/30
        return [self._python_rpath, '-m', 'pip']

    @property
    def _python_rpath(self):
        """The relative path (from environment root) to python."""
        # Windows virtualenv installation installs pip to the [Ss]cripts
        # folder. Here's a simple check to support:
        if sys.platform == 'win32':
            return os.path.join('Scripts', 'python.exe')
        return os.path.join('bin', 'python')

    @property
    def pip_version(self):
        """Version of installed pip."""
        if not self._pip_exists:
            return None
        if not hasattr(self, '_pip_version'):
            # don't call `self._execute_pip` here as that method calls this one
            output = self._execute(self._pip + ['-V'], log=False).split()[1]
            self._pip_version = tuple([int(n) for n in output.split('.')])
        return self._pip_version

    @property
    def root(self):
        """The root directory that this virtual environment exists in."""
        return os.path.split(self.path)[0]

    @property
    def name(self):
        """The name of this virtual environment (taken from its path)."""
        return os.path.basename(self.path)

    @property
    def _logfile(self):
        """Absolute path of the log file for recording installation output."""
        return os.path.join(self.path, 'build.log')

    @property
    def _errorfile(self):
        """Absolute path of the log file for recording installation errors."""
        return os.path.join(self.path, 'build.err')

    def _create(self):
        """Executes `virtualenv` to create a new environment."""
        if self.readonly:
            raise VirtualenvReadonlyException()
        args = ['virtualenv']
        if self.system_site_packages:
            args.append('--system-site-packages')
        if self.python is None:
            args.append(self.name)
        else:
            args.extend(['-p', self.python, self.name])
        proc = subprocess.Popen(args, cwd=self.root, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
        output, error = proc.communicate()
        returncode = proc.returncode
        if returncode:
            raise VirtualenvCreationException((returncode, output, self.name))
        self._write_to_log(output, truncate=True)
        self._write_to_error(error, truncate=True)

    def _execute_pip(self, args, log=True):
        """
        Executes pip commands.

        :param args: Arguments to pass to pip (list[str])
        :param log: Log the output to a file [default: True] (boolean)
        :return: See _execute
        """

        # Copy the pip calling arguments so they can be extended
        exec_args = list(self._pip)

        # Older versions of pip don't support the version check argument.
        # Fixes https://github.com/sjkingo/virtualenv-api/issues/35
        if self.pip_version[0] >= 6:
            exec_args.append('--disable-pip-version-check')

        exec_args.extend(args)
        return self._execute(exec_args, log=log)

    def _execute(self, args, log=True):
        """Executes the given command inside the environment and returns the output."""
        if not self._ready:
            self.open_or_create()
        output = ''
        error = ''
        try:
            proc = subprocess.Popen(args, cwd=self.path, env=self.env, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
            output, error = proc.communicate()
            returncode = proc.returncode
            if returncode:
                raise subprocess.CalledProcessError(returncode, proc, (output, error))
            return to_text(output)
        except OSError as e:
            # raise a more meaningful error with the program name
            prog = args[0]
            if prog[0] != os.sep:
                prog = os.path.join(self.path, prog)
            raise OSError('%s: %s' % (prog, six.u(str(e))))
        except subprocess.CalledProcessError as e:
            output, error = e.output
            e.output = output
            raise e
        finally:
            if log:
                try:
                    self._write_to_log(to_text(output))
                    self._write_to_error(to_text(error))
                except NameError:
                    pass  # We tried

    def _write_to_log(self, s, truncate=False):
        """Writes the given output to the log file, appending unless `truncate` is True."""
        # if truncate is True, set write mode to truncate
        with open(self._logfile, 'w' if truncate else 'a') as fp:
            fp.writelines((to_text(s) if six.PY2 else to_text(s), ))

    def _write_to_error(self, s, truncate=False):
        """Writes the given output to the error file, appending unless `truncate` is True."""
        # if truncate is True, set write mode to truncate
        with open(self._errorfile, 'w' if truncate else 'a') as fp:
            fp.writelines((to_text(s)), )

    def _pip_exists(self):
        """Returns True if pip exists inside the virtual environment. Can be
        used as a naive way to verify that the environment is installed."""
        return os.path.isfile(os.path.join(self.path, 'bin', 'pip'))

    def open_or_create(self):
        """Attempts to open the virtual environment or creates it if it
        doesn't exist.
        XXX this should probably be expanded to do some proper checking?"""
        if not self._pip_exists():
            self._create()
        self._ready = True

    def install(self, package, force=False, upgrade=False, options=None):
        """Installs the given package into this virtual environment, as
        specified in pip's package syntax or a tuple of ('name', 'ver'),
        only if it is not already installed. Some valid examples:

         'Django'
         'Django==1.5'
         ('Django', '1.5')
         '-e .'
         '-r requirements.txt'

        If `force` is True, force an installation. If `upgrade` is True,
        attempt to upgrade the package in question. If both `force` and
        `upgrade` are True, reinstall the package and its dependencies.
        The `options` is a list of strings that can be used to pass to
        pip."""
        if self.readonly:
            raise VirtualenvReadonlyException()
        if options is None:
            options = []
        if isinstance(package, tuple):
            package = '=='.join(package)
        if package.startswith(('-e', '-r')):
            package_args = package.split()
        else:
            package_args = [package]
        if not (force or upgrade) and (package_args[0] != '-r' and self.is_installed(package_args[-1])):
            self._write_to_log('%s is already installed, skipping (use force=True to override)' % package_args[-1])
            return
        if not isinstance(options, list):
            raise ValueError("Options must be a list of strings.")
        if upgrade:
            options += ['--upgrade']
            if force:
                options += ['--force-reinstall']
        elif force:
            options += ['--ignore-installed']
        try:
            self._execute_pip(['install'] + package_args + options)
        except subprocess.CalledProcessError as e:
            raise PackageInstallationException((e.returncode, e.output, package))

    def uninstall(self, package):
        """Uninstalls the given package (given in pip's package syntax or a tuple of
        ('name', 'ver')) from this virtual environment."""
        if isinstance(package, tuple):
            package = '=='.join(package)
        if not self.is_installed(package):
            self._write_to_log('%s is not installed, skipping' % package)
            return
        try:
            self._execute_pip(['uninstall', '-y', package])
        except subprocess.CalledProcessError as e:
            raise PackageRemovalException((e.returncode, e.output, package))

    def wheel(self, package, options=None):
        """Creates a wheel of the given package from this virtual environment,
        as specified in pip's package syntax or a tuple of ('name', 'ver'),
        only if it is not already installed. Some valid examples:

         'Django'
         'Django==1.5'
         ('Django', '1.5')

        The `options` is a list of strings that can be used to pass to
        pip."""
        if self.readonly:
            raise VirtualenvReadonlyException()
        if options is None:
            options = []
        if isinstance(package, tuple):
            package = '=='.join(package)
        if not self.is_installed('wheel'):
            raise PackageWheelException((0, "Wheel package must be installed in the virtual environment", package))
        if not isinstance(options, list):
            raise ValueError("Options must be a list of strings.")
        try:
            self._execute_pip(['wheel', package] + options)
        except subprocess.CalledProcessError as e:
            raise PackageWheelException((e.returncode, e.output, package))

    def is_installed(self, package):
        """Returns True if the given package (given in pip's package syntax or a
        tuple of ('name', 'ver')) is installed in the virtual environment."""
        if isinstance(package, tuple):
            package = '=='.join(package)
        if package.endswith('.git'):
            pkg_name = os.path.split(package)[1][:-4]
            return pkg_name in self.installed_package_names or \
                    pkg_name.replace('_', '-') in self.installed_package_names
        pkg_tuple = split_package_name(package)
        if pkg_tuple[1] is not None:
            return pkg_tuple in self.installed_packages
        else:
            return pkg_tuple[0].lower() in self.installed_package_names

    def upgrade(self, package, force=False):
        """Shortcut method to upgrade a package. If `force` is set to True,
        the package and all of its dependencies will be reinstalled, otherwise
        if the package is up to date, this command is a no-op."""
        self.install(package, upgrade=True, force=force)

    def upgrade_all(self):
        """
        Upgrades all installed packages to their latest versions.
        """
        for pkg in self.installed_package_names:
            self.install(pkg, upgrade=True)

    def search(self, term):
        """
        Searches the PyPi repository for the given `term` and returns a
        dictionary of results.

        New in 2.1.5: returns a dictionary instead of list of tuples
        """
        packages = {}
        results = self._execute_pip(['search', term], log=False)  # Don't want to log searches
        for result in results.split(linesep):
            try:
                name, description = result.split(six.u(' - '), 1)
            except ValueError:
                # '-' not in result so unable to split into tuple;
                # this could be from a multi-line description
                continue
            else:
                name = name.strip()
                if len(name) == 0:
                    continue
                packages[name] = description.split(six.u('<br'), 1)[0].strip()
        return packages

    def search_names(self, term):
        return list(self.search(term).keys())

    @property
    def installed_packages(self):
        """
        List of all packages that are installed in this environment in
        the format [(name, ver), ..].
        """
        freeze_options = ['-l', '--all'] if self.pip_version >= (8, 1, 0) else ['-l']
        return list(map(split_package_name, filter(None, self._execute_pip(
                ['freeze'] + freeze_options).split(linesep))))

    @property
    def installed_package_names(self):
        """List of all package names that are installed in this environment."""
        return [name.lower() for name, _ in self.installed_packages]
